Fathead for NumPy

ashishbhate opened the pull request:
## Description of new Instant Answer, or changes
<!-- What does this new Instant Answer do? What changes does this PR introduce? -->
NumPy fathead

## Related Issues and Discussions
Fixes #233 

## People to notify
<!-- Please @mention any relevant people/organizations here: -->
@pjhampton

<!-- DO NOT REMOVE -->
---


<!-- The Instant Answer ID can be found by clicking the `?` icon beside the Instant Answer result on DuckDuckGo.com -->
Instant Answer Page: https://duck.co/ia/view/numpy
<!-- FILL THIS IN:                           ^^^^ -->

## Notes:
I'm not entirely sure I've got the escaping right.

daxtheduck commented:
@ashishbhate Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

moollaza commented:
@ashishbhate excellent work 👍 💯 

This is now live on Beta: https://beta.duckduckgo.com/?q=numpy+chararray+index&t=h_&ia=about

I think this may already be mergeable 😄

moollaza commented on lib/fathead/numpy/parse.py:
```diff
@@ -0,0 +1,344 @@
+    rendered_information = INFORMATION_TEMPLATE.format(information=information)
+
+    raw_signature = [each.string for each in section_div.dt.children]
+    signature = ''.join(raw_signature).strip().replace("¶", "")
```
We should remove `\t` here instead of escaping it later. I don't think `\t` is necessary for he description text.

We may also want to remove the `\n` -- I'm not sure how much of a difference they make?

moollaza replied:
@ashishbhate it turns out we need to replace `\t` with four spaces. Our internal system throws out articles that have `\\t` in the abstract, but if we put 4 spaces in it's place we'll be fine 👍

moollaza commented:
@ashishbhate it looks like we always have a space, then a newline before the closing `</code>`?

Let's remove the space and `\n` as they're not necessary 👍

pjhampton commented:
This is awesome 😄  Thanks for this @ashishbhate! I used numpy daily so this will come in really handy. There is a couple of things I can't get to trigger such as `numpy loadtxt` or `numpy random seed`

Also, I'm sure you and @moollaza, people tend to refer to numpy as just **np**, when I search I type np, not numpy. Maybe this could be a potential alias?

moollaza commented:
@ashishbhate if you push the `output.txt` to the PR, this will now auto install to Beta 👍 

As well if you add a `trigger_words.txt` file, we can automatically add some triggers for this Fathead. For example:

```
#trigger_words.txt
numpy
np
```

With this, we can remove `numpy` from the article titles and then we can allow `numpy` to be at the start OR end of the query. The same will also apply for `np`.

I can force the Fathead to require the presence of one of those triggers if the article titles are very generic, by checking off the `isFanon` option on the IA Page.

ashishbhate commented:
@moollaza I just wanted to confirm it's ok to push the `output.txt` file. The documentation says not to push the file if its greater than 1MB. The file generated by this fathead processor is about 1.5MB

ashishbhate commented:
@moollaza @pjhampton Thanks for the feedback! 

I don't think I'll get around to addressing the review comments this week. I hope to have it done by next week.

- [ ] Remove whitespace and newline at the end of `</code>`
- [ ] Replace `\t` with 4 spaces
- [ ] Include a `trigger_words.txt` files
- [ ] Update redirects (Add a leading `numpy` if title doesn't have one)
- [ ] Include `output.txt` file

moollaza commented:
> @moollaza I just wanted to confirm it's ok to push the output.txt file. The documentation says not to push the file if its greater than 1MB. The file generated by this fathead processor is about 1.5MB

@ashishbhate yes it's fine, 1.5MB is okay. We just don't want anything really big. For the purpose of auto-deploying you'll need to push your output.txt anyways. It can always be removed in the PR and then it will not affect the repo as the commits are squashed.

moollaza commented:
@ashishbhate Ping! Would you be able to finish this one up? We'd love to get this live!

ashishbhate commented:
@moollaza I'll definitely get to it in a day or two.

ashishbhate commented:
@pjhampton @moollaza Updated. A few things:

1) Entries with parameters and returns are now correctly triggered. Try [numpy.zeros](https://beta.duckduckgo.com/?q=numpy.zeros&t=h_&ia=about).
`numpy.zeros` triggers correctly but `np.zeros` doesn't. I've added a `trigger_words.txt` file as you suggested. Do I also need to add a `np.foo.bar` redirect in the `output.txt` file for  the `trigger_words.txt` entry to take effect?

2) I'm not sure that I should display the parameters and returns in `<code>` tags. DDG attempts to do syntax highlighting which makes the text look a bit strange.

jdorweiler commented:
Could you include code examples if they exist on the page? I saw that the page for 'numpy.array' includes some examples at the end. 
https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html

ashishbhate commented:
Sorry about the delayed response. I've been afk most of the last two weeks on account of the holidays.

I haven't added examples because even without examples some of the IA responses are quite long. 
By adding the examples we'd end up reproducing nearly the whole documentation page in the IA response (atleast in a lot of the cases). 

I'm not opposed to adding the examples. I was under the impression that the IA response was supposed to be a quick reference rather than a complete reproduction of documentation content, with a link at the bottom that points to the documentation page. 

@moollaza could you weigh in on this? 

Also, gently bumping this.

moollaza commented:
> @moollaza could you weigh in on this?

I think what @jdorweiler meant was to add perhaps one example usage, not necessarily all the examples in the docs.

I agree that copy over all the examples would be overkill. I personally prefer to see an example usage along with the result to help me more clearly understand how a function/method is used and what it returns. Many of our Fatheads show an example usage, followed by a description and explanation.

daxtheduck commented:
@ashishbhate this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

moollaza commented:
@ashishbhate In any case, this is more than good enough to be merged. We can add examples in another PR, or create a task for someone else to do so, if you'd like.

Thanks again! This LGTM :+1:

ashishbhate commented:
Thanks @moollaza!

I'll start work on the SciPy fathead which will probably be nearly the same code with a few URL changes. 

There are two minor issues with this fathead. I've listed them [here](https://github.com/duckduckgo/zeroclickinfo-fathead/pull/584#issuecomment-267319482). I can fix them once I have your feedback.

I've opened an issue for adding examples: #684 (someone will need to label that issue).  /cc @jdorweiler

moollaza commented:
> but np.zeros doesn't. I've added a trigger_words.txt file as you suggested. Do I also need to add a np.foo.bar redirect in the output.txt file for the trigger_words.txt entry to take effect?

@ashishbhate `np.zero` will only work if you create a redirect for it. Internally we strip `.` from queries, so we ultimately see `npzeros`. With `np` as a trigger the query `np zeros` should trigger though.

Trigger words must be explicitly separated by a space.

> I'm not sure that I should display the parameters and returns in <code> tags. DDG attempts to do syntax highlighting which makes the text look a bit strange.

Yes, let's move that out of a code black. We can use a `<ul>` instead. A table might also work...
